watchtower-plugin/src/wt_client.rs

```diff
@@ -0,0 +1,420 @@
+use std::collections::HashMap;
+use std::path::PathBuf;
+use std::sync::{Arc, Mutex};
+use tokio::fs;
+
+use bitcoin::secp256k1::{PublicKey, Secp256k1, SecretKey};
+
+use teos_common::appointment::{Appointment, Locator};
+use teos_common::cryptography;
+use teos_common::dbm::Error as DBError;
+use teos_common::receipts::{AppointmentReceipt, RegistrationReceipt};
+use teos_common::{UserId, UserId as TowerId};
+
+use crate::dbm::DBM;
+use crate::{MisbehaviorProof, TowerInfo, TowerStatus, TowerSummary};
+
+/// Represents the watchtower client that is being used as the CoreLN plugin state.
+#[derive(Clone)]
+pub struct WTClient {
+    /// A [DBM] instance.
+    pub dbm: Arc<Mutex<DBM>>,
+    /// A collection of towers the client is registered to.
+    pub towers: HashMap<TowerId, TowerSummary>,
+    /// The user secret key.
+    pub user_sk: SecretKey,
+    /// The user identifier.
+    pub user_id: UserId,
+}
+
+impl WTClient {
+    pub async fn new(data_dir: PathBuf) -> Self {
+        // Create data dir if it does not exist
+        fs::create_dir_all(&data_dir).await.unwrap_or_else(|e| {
+            log::error!("Cannot create data dir: {:?}", e);
+            std::process::exit(1);
+        });
+
+        let dbm = DBM::new(&data_dir.join("watchtowers_db.sql3")).unwrap();
+        let (user_sk, user_id) = match dbm.load_client_key() {
+            Ok(sk) => (
+                sk,
+                UserId(PublicKey::from_secret_key(&Secp256k1::new(), &sk)),
+            ),
+            Err(_) => {
+                log::info!("Watchtower client keys not found. Creating a fresh set");
+                let (sk, pk) = cryptography::get_random_keypair();
+                dbm.store_client_key(&sk).unwrap();
+                (sk, UserId(pk))
+            }
+        };
+
+        log::info!(
+            "Plugin watchtower client initialized. User id = {}",
+            user_id
+        );
+
+        WTClient {
+            towers: dbm.load_towers(),
+            dbm: Arc::new(Mutex::new(dbm)),
+            user_sk,
+            user_id,
+        }
+    }
+
+    /// Adds or updates a tower entry.
+    pub fn add_update_tower(
+        &mut self,
+        tower_id: TowerId,
+        tower_net_addr: String,
+        receipt: &RegistrationReceipt,
+    ) {
+        self.dbm
+            .lock()
+            .unwrap()
+            .store_tower_record(tower_id, &tower_net_addr, receipt)
+            .unwrap();
+        self.towers.insert(
+            tower_id,
+            TowerSummary::new(
+                tower_net_addr,
+                receipt.available_slots(),
+                receipt.subscription_expiry(),
+            ),
+        );
+    }
+
+    /// Loads a tower record from the database.
+    pub fn load_tower_info(&self, tower_id: TowerId) -> Result<TowerInfo, DBError> {
+        self.dbm.lock().unwrap().load_tower_record(tower_id)
```

Comment:
After loading the tower record, we should patch the true tower status from `self.towers` since `dbm.load_tower_record()` will always say that the tower is reachable.

Reply:
That's actually fixed in the retry PR:

https://github.com/talaia-labs/rust-teos/blob/10c6b65bea83dce5737fc8abeddd142afb3a5431/watchtower-plugin/src/dbm.rs#L192-L193

https://github.com/talaia-labs/rust-teos/blob/10c6b65bea83dce5737fc8abeddd142afb3a5431/watchtower-plugin/src/dbm.rs#L223-L225

But I can move it to the base PR if you think it makes more sense to be here.